### ISSUE TITLE
Document sources of deliberate unsoundness

### ISSUE THREAD

**Issue by Richargh** [uber/NullAway#1617]
## Description

NullAway currently does not warn when doing nested null-checks with mutable data. I can write: 

```java
if(rental.returnPolicy() == null || rental.returnPolicy().mustBeReturnedAfter() == null){
    //                                                                                   ^^^^^
    // is not, but should be: [NullAway] dereferenced expression rental.returnPolicy() is @Nullable
    // returnPolicy is just a method so its return value could change between first and second call
    return 0;
}
```

To me that seems like a false negative. It does not warn but should. 

## Minimal repro

```java
import org.jetbrains.annotations.Nullable;

public class Main {
    public static void main(String[] args){
        check(new RegularRental());
    }

    static int check(Rental rental){
        if(rental.returnPolicy() == null || rental.returnPolicy().mustBeReturnedAfter() == null){
            //                                                    ^^^^^^^^^^^^^^^^^^^^
            // is not, but should be: [NullAway] dereferenced expression rental.returnPolicy() is @Nullable
            // returnPolicy is just a method so its return value could change between first and second call
            return 0;
        }
        return 1;
    }
}


interface Rental {
    @Nullable ReturnPolicy returnPolicy();
}

class RegularRental implements Rental {

    private ReturnPolicy returnPolicy = new ReturnPolicy("2");

    @Override public ReturnPolicy returnPolicy(){ return this.returnPolicy; }
}

class ReturnPolicy {
    private @Nullable String mustBeReturnedAfter;

    ReturnPolicy(@Nullable String mustBeReturnedAfter) { this.mustBeReturnedAfter = mustBeReturnedAfter; }

    public @Nullable String mustBeReturnedAfter() { return mustBeReturnedAfter; }
}
```

**Comment by msridhar** [comment]
@Richargh thanks for the report!  This is actually a deliberate false negative.  NullAway assumes all methods are pure, i.e., free of side effects and deterministic.  In this case, it assumes that `rental.returnPolicy()` returns the same value each time it is called, which I agree could lead to a missed issue.  This decision was made for practicality, since it reduces the number of annotations developers need to write.  See Section 4 of [this paper](https://manu.sridharan.net/files/FSE19NullAway.pdf) for more details.  If you are looking for a checker that more soundly handles cases like this, I suggest the [Nullness Checker](https://checkerframework.org/manual/#nullness-checker) from the Checker Framework.

We should probably write a wiki page better documenting sources of deliberate unsoundness in NullAway; I'll leave this issue open to track that.

**Comment by Andrei-hub11** [comment]
@msridhar, I'd like to work on this. Could you please assign it to me?

**Comment by msridhar** [comment]
@Andrei-hub11 done!  You can't make a pull request directly on the wiki, unfortunately.  But, you can make a PR on this repo, and then I'll sync with the main wiki:

https://github.com/msridhar/NullAway-wiki

I'd suggest a new top-level page on this topic.

**Comment by Richargh** [comment]
Hey @msridhar thank you for explaining this.

The documentation would be very much appreciated. I love your tool and it's very good to know what the intentional limits are. (Full disclosure, I featured NullAway in an article on ["Fixing the Billion-Dollar Mistake"](https://javapro.io/2025/08/07/fixing-the-billion-dollar-mistake/) and am right now writing another ["Kotlin kontra Java"](https://javapro.io/?s=kotlin+kontra+java) article that focuses on (null-)safe code)

For this case I don't understand the reasoning though. Why would it require more annotations? I can see why it would require more code, I added the code that would be necessary below, but why more annotations? 

```java
var policy = rental.returnPolicy();
var mustBeReturnedAfter = policy != null
    ? policy.mustBeReturnedAfter()
    : null;
if(mustBeReturnedAfter == null){
    return 0;
}
```

**Comment by msridhar** [comment]
> The documentation would be very much appreciated. I love your tool and it's very good to know what the intentional limits are. (Full disclosure, I featured NullAway in an article on ["Fixing the Billion-Dollar Mistake"](https://javapro.io/2025/08/07/fixing-the-billion-dollar-mistake/) and am right now writing another ["Kotlin kontra Java"](https://javapro.io/?s=kotlin+kontra+java) article that focuses on (null-)safe code)

https://github.com/msridhar/NullAway-wiki/pull/3 should help improve the docs, once I can make a pass over it and land it.  Thanks for the great articles!

> 
> For this case I don't understand the reasoning though. Why would it require more annotations? I can see why it would require more code, I added the code that would be necessary below, but why more annotations?
> ```java
> var policy = rental.returnPolicy();
> var mustBeReturnedAfter = policy != null
>     ? policy.mustBeReturnedAfter()
>     : null;
> if(mustBeReturnedAfter == null){
>     return 0;
> }
> ```

So, the full explanation is a bit complex.  First, let's assume the program is single threaded.  Then, we have two cases:

1. `returnPolicy()` is deterministic: in this case, even the original code is free of NPEs, since `rental.returnPolicy()` is guaranteed to return the same value each time it's called.
2. `returnPolicy()` is non-deterministic: in this case, the `rental.returnPolicy()` calls might return non-null the first time and null the second time.  For this case, your modified code that adds the `policy` local variable actually could behave differently than the original program; the second `rental.returnPolicy()` might not get the "latest" return value of the function.  Probably in most cases adding the local variable is fine, but at least hypothetically it might not be what the programmer wants.

In any case, Checker Framework has a [`@Deterministic`](https://checkerframework.org/api/org/checkerframework/dataflow/qual/Deterministic.html) annotation you can put on methods to document they are deterministic, but NullAway does not support this and just assumes determinism.

Now, for your full code example, I think the actual implementation of `RegularRental.returnPolicy()` could only return different values for the two calls you wrote if there is multithreading and the `RegularRental.returnPolicy` field gets overwritten by another thread between the calls.  But, even the Checker Framework Nullness Checker is unsound in the presence of multithreading, at least by default.  E.g., the Nullness Checker does not report any warnings for this code, though it could NPE if another thread can overwrite `w.f` between the two reads:

```java
class Test {
    static class Wrapper {
        @Nullable Wrapper f;
    }
    void test(Wrapper w) {
        if (w.f != null && w.f.f != null) {
            w.f.f.hashCode();
        }
    }
}
```
You can similarly add a local variable to make this example robust against NPEs, but with the potential change to program behavior in corner cases.

**Comment by msridhar** [comment]
I have merged @Andrei-hub11's PR, so there is now a wiki page documenting these issues:

https://github.com/uber/NullAway/wiki/Deliberate-Unsoundness

Thanks!
